core/src/value/deserialize.rs

```diff
@@ -29,55 +50,286 @@ pub enum Error {
+     * If structures contain invalid combination of date/time components, it fails.
+     */
+    fn latest(&self) -> Result<U>;
+
+    /**
+     * Returns a tuple of the earliest and latest possible value from a partial precision structure.
+     * 
+     */
+    fn to_range(&self) -> Result<(Option<U>, Option<U>)> {
+        if self.is_precise() {
+            ZeroRange.fail()
+        } else {
+            Ok((self.earliest().ok(), self.latest().ok()))
+        }
+    }
+
+    /**
+     * Returns true, if partial precision structure has maximum possible accuracy.
+     * For fraction of a second, ... TODO decide what to do, as RUST Chrono is more
+     * precise than 6 digits of DICOM standard ...
+     */
+    fn is_precise(&self) -> bool {
+        let e = self.earliest();
+        let l = self.latest();
+
+        e.is_ok() && l.is_ok() && e.ok() == l.ok()
+    }
+}
+
+impl Imprecision<NaiveDate> for PartialDate {
+    fn earliest(&self) -> Result<NaiveDate> {
+        let (y, m, d) = match self.components {
+            DatePrecision::Year(y) => (y, 1, 1),
+            DatePrecision::Month(y, m) => (y, m, 1),
+            DatePrecision::Day(y, m, d) => (y, m, d),
+        };
+        let date: Result<_> =
+            NaiveDate::from_ymd_opt(y as i32, m as u32, d as u32).context(InvalidDate);
+        Ok(date?)
+    }
+
+    fn latest(&self) -> Result<NaiveDate> {
+        let (y, m, d) = match self.components {
+            DatePrecision::Year(y) => (y, 12, 31),
+            DatePrecision::Month(y, m) => {
+                let d = {
+                    if m == 12 {
+                        NaiveDate::from_ymd(y as i32 + 1, 1, 1) 
+                    }
+                    else{
+                        NaiveDate::from_ymd(y as i32, m as u32 + 1, 1)
+                    }.signed_duration_since(NaiveDate::from_ymd(y as i32, m as u32, 1))
+                    .num_days()
+                };
+                (y, m, d as u8)
+            },
+            DatePrecision::Day(y, m, d) => (y, m, d),
+        };
+        let date: Result<_> =
+            NaiveDate::from_ymd_opt(y as i32, m as u32, d as u32).context(InvalidDate);
+        Ok(date?)
+    }
+}
+
+fn check_component(component: DateComponent, value: &u32) -> Result<()> {
+    let range = match component {
+        DateComponent::Year => 0..=9_999,
+        DateComponent::Month => 1..=12,
+        DateComponent::Day => 1..=31,
+        DateComponent::Hour => 0..=23,
+        DateComponent::Minute => 0..=59,
+        DateComponent::Second => 0..=59,
+        DateComponent::Fraction => 0..=1_999_999,
+        //is not i32 for simplification,  so only check UTC range, not sign
+        DateComponent::UTCOffset => 0..=86_399,
+    };
+
+    if range.contains(value) {
+        Ok(())
+    } else {
+        InvalidComponent {
+            component: component,
+            value: value.clone(),
+            range: range,
```

Comment:
Running `cargo clippy` every once in a while will give you tips to make the code tidier.

```suggestion
            component,
            value,
            range,
```